Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Internal/CoreRemoteMongoCollection.swift

```diff
@@ -99,6 +98,40 @@ public class CoreRemoteMongoCollection<T: Codable> {
         return CoreRemoteMongoReadOperation<CollectionType>.init(command: "find", args: args, service: self.service)
     }
 
+    /**
+     * Finds a document in this collection which matches the provided filter.
+     *
+     * - parameters:
+     *   - filter: A `Document` that should match the query.
+     *   - options: Optional `RemoteFindOptions` to use when executing the command.
+     *
+     * - returns: The resulting `Document` or nil if no such document exists
+     */
+    public func findOne(_ filter: Document = [:], options: RemoteFindOptions? = nil) throws -> T? {
+        var args = baseOperationArgs
+
+        args["query"] = filter
+        if let options = options {
+            if let projection = options.projection {
+                args[RemoteFindOptionsKeys.projection.rawValue] = projection
+            }
+            if let sort = options.sort {
+                args[RemoteFindOptionsKeys.sort.rawValue] = sort
+            }
+        }
+
```

Comment:
Below is definitely not the prettiest way to go about this, but I was not sure how / if we want to change the implementation of CallFunction in the case of the request returning nil. Simply catching seemed like the least invasive way to go about this, but let me know if there is a better way to go about this

Reply:
We definitely don't want to be parsing error message strings for a common type of result. We should probably add another version of `doAuthenticatedRequest` that returns an optional T, and the compiler should automatically pick that one in `findOne`.

It would look something like

```swift
    public func doAuthenticatedRequest<T: Decodable>(_ stitchReq: StitchAuthRequest) throws -> T? {
        let response = try self.doAuthenticatedRequest(stitchReq)
        do {
            guard let responseBody = response.body,
                let responseString = String.init(data: responseBody, encoding: .utf8) else {
                    throw StitchError.serviceError(
                        withMessage: StitchErrorCodable.genericErrorMessage(withStatusCode: response.statusCode),
                        withServiceErrorCode: .unknown)
            }

            do {
                return try BSONDecoder().decode(T?.self, from: responseString)
            } catch let err {
                throw StitchError.requestError(withError: err, withRequestErrorCode: .decodingError)
            }
        }
    }
```

assuming that BSON undefined can be decoded to nil. I just checked and it compiles, but it should be tested. If that doesn't work, we should discuss the best approach with the Swift driver team.